components/support/sql/src/open_database.rs

```diff
@@ -0,0 +1,574 @@
+
+    // The first version that this migration applies to (usually 1)
+    const START_VERSION: u32;
+
+    // The version that the last upgrade function upgrades to.
+    const END_VERSION: u32;
+
+    // Runs before the init/upgrade functions
+    fn prepare(&self, _conn: &Connection) -> Result<()> {
+        Ok(())
+    }
+
+    // Initialize a newly created database to END_VERSION
+    fn init(&self, conn: &Connection) -> Result<()>;
+
+    // Upgrade schema from version -> version + 1
+    fn upgrade_from(&self, conn: &Connection, version: u32) -> Result<()>;
+
+    // Runs after the init/upgrade functions
+    fn finish(&self, _conn: &Connection) -> Result<()> {
+        Ok(())
+    }
+}
+
+fn run_migration_logic<ML: MigrationLogic>(
+    migration_logic: &ML,
+    conn: &Connection,
+    init: bool,
+) -> Result<()> {
+    log::debug!("{}: opening database", ML::NAME);
+    let tx = conn.unchecked_transaction()?;
+    log::debug!("{}: preparing database", ML::NAME);
+    migration_logic.prepare(&tx)?;
+    if init {
+        log::debug!("{}: initializing new database", ML::NAME);
+        migration_logic.init(&tx)?;
+    } else {
+        let mut current_version = get_schema_version(&tx)?;
+        if current_version < ML::START_VERSION {
+            return Err(Error::VersionTooOld(current_version));
+        } else if current_version > ML::END_VERSION {
+            return Err(Error::VersionTooNew(current_version));
+        }
+        while current_version < ML::END_VERSION {
+            log::debug!(
+                "{}: upgrading database to {}",
+                ML::NAME,
+                current_version + 1
+            );
+            migration_logic.upgrade_from(&tx, current_version)?;
+            current_version += 1;
+        }
+    }
+    log::debug!("{}: finishing database open", ML::NAME);
+    migration_logic.finish(&tx)?;
+    set_schema_version(&tx, ML::END_VERSION)?;
+    tx.commit()?;
+    log::debug!("{}: database open successful", ML::NAME);
+    Ok(())
+}
+
+pub fn open_database<ML: MigrationLogic>(
+    path: PathBuf,
+    migration_logic: &ML,
+    error_handling: ErrorHandling,
+) -> Result<Connection> {
+    open_database_with_flags(
+        DatabaseLocation::File(path),
+        OpenFlags::default(),
+        migration_logic,
+        error_handling,
+    )
+}
+
+pub fn open_memory_database<ML: MigrationLogic>(
+    migration_logic: &ML,
+    error_handling: ErrorHandling,
+) -> Result<Connection> {
+    open_database_with_flags(
+        DatabaseLocation::Memory,
+        OpenFlags::default(),
+        migration_logic,
+        error_handling,
+    )
+}
+
+pub fn open_database_with_flags<ML: MigrationLogic>(
+    location: DatabaseLocation,
+    open_flags: OpenFlags,
+    migration_logic: &ML,
+    error_handling: ErrorHandling,
+) -> Result<Connection> {
+    // Try running the migration logic with an existing file
+    let initializing = !location.exists();
+    let mut conn = location.open(open_flags)?;
```

Comment:
Errors here should probably have the `DeleteAndRecreate` treatment? I'm thinking of corrupted databases mainly.